readthedocs/projects/tasks.py

```diff
@@ -1005,3 +1007,26 @@ def sync_callback(_, version_pk, commit, *args, **kwargs):
     """
     fileify(version_pk, commit=commit)
     update_search(version_pk, commit=commit)
+
+
+@app.task()
+def finish_inactive_builds():
+    delta = datetime.timedelta(hours=1, minutes=30)
+    query = (~Q(state=BUILD_STATE_FINISHED) &
+             Q(date__lte=datetime.datetime.now() - delta))
+
+    builds = Build.objects.filter(query)[:50]
+    for build in builds:
+        build.success = False
```

Comment:
Let's check for the timeout date here, ie - `if build.date + build.project.container_time_limit > now: continue`.

So the query will surface extra builds, but the custom timeout builds will be an exception to this case. They keep running through this loop until the condition above is true.

Does this seem reasonable?

Reply:
Yes. I think this is the easiest and more reasonable thing to do.

"I was overengeering it" as Eric said :)